test/compile/selection/scales.test.ts

```diff
@@ -36,11 +37,11 @@ describe('Selection + Scales', function() {
             },
             color: {
               field: "symbol", type: "nominal",
-              scale: {domain: {selection: "brush2"}}
+              scale: {domain: {selection: "brush2"} as Domain}
```

Comment:
I think it was previously not detected by Typescript because we didn't require `value` in ValueDef and `type` in FieldDef -- thus preventing it from infer if a spec is a unit spec and so on.  

In this ambiguous state, somehow Typescript just ignore type checking totally.  I guess this behavior makes sense for a codebase that migrates from JS and thus partially have type defined.  

Anyway, I'm glad I notice this and have fixed a lot of things in this PR :)